### ISSUE TITLE
[BOLT] BOLT counts x86-64 General-Dynamic / Local-Dynamic TLS relocations as "Failed to analyze"

### ISSUE THREAD

**Issue by not4juu** [llvm/llvm-project#213037]
## Description

It's just rather minor problem does not affect the executing code but just wrong reporting from BOLT which does not follow already-existing TLS relocation logic inside BOLT during  `llvm-bolt` usage on x86-64 lib with option `--emit-relocs`, BOLT reports:

```
BOLT-WARNING: Failed to analyze N relocations
```

for **General-Dynamic (GD)** and **Local-Dynamic (LD)** TLS relocations, even though these relocations are harmless for BOLT (they encode TLS offsets / module IDs, not code or data addresses that BOLT relocates).

Root cause: `isTLSX86()` from `bolt/lib/Core/Relocation.cpp` only classifies the **Initial-Exec / Local-Exec** TLS relocations as TLS. GD/LD relocations are therefore *not* recognized as TLS, skip the "no special handling for TLS on x86" early-out in `handleRelocation()`, fall through to `analyzeRelocation()`, fail the `isSupported()` check and get counted as failures. The warning is noisy and misleading (it suggests BOLT could not understand the binary), and on large real-world PIC shared objects the reported count N reaches into the thousands.

> Note the inconsistency with the other targets: on `AArch64` the dynamic `TLS` model (`TLSDESC`, the `AArch64` analogue of `x86 GD/LD`) is already classified as `TLS` by `isTLSAArch64()`.

## Environment

- Target: `x86_64`, PIC shared object / PIE (relocation mode, `--emit-relocs`).
- Reproduced on current `main` (`llvm-bolt` version: bd6adfedc776c07caf158e59367d9c246c933510).
- Compiler: any (`g++ 14.2`, `clang++` - from the same LLVM revision as the `llvm-bolt` above).

### reproducer.cpp
``` c++
// --- General-Dynamic (GD): default-visibility, preemptible thread_local ------
// In a shared object this stays GD (no linker relaxation) -> R_X86_64_TLSGD.
__attribute__((tls_model("global-dynamic"))) thread_local long gd_var = 1;

long read_gd()
{
    return gd_var;
}

// --- Local-Dynamic (LD): several hidden/local thread_local vars accessed -----
// together in one function -> one R_X86_64_TLSLD (module base) plus one
// R_X86_64_DTPOFF32 per variable. The runtime `delta` argument prevents the
// optimizer from constant-folding the accesses away.
static __attribute__((tls_model("local-dynamic"))) thread_local long ld_a = 2;
static __attribute__((tls_model("local-dynamic"))) thread_local long ld_b = 3;

long read_ld(long delta)
{
    ld_a += delta;
    ld_b += delta;
    return ld_a + ld_b;
}
```

Build (PIC shared object, keep relocations):

```sh
g++ -O2 -fPIC -shared -fno-plt -Wl,--emit-relocs reproducer.cpp -o libGcc.so
clang++ -O2 -fPIC -shared -fno-plt -Wl,--emit-relocs reproducer.cpp -o libClang.so
```

The code (`.rela.text`) relocations produced check

```
$ readelf -rW libGcc.so | grep -E 'TLSGD|TLSLD|DTPOFF32'
... R_X86_64_TLSGD     gd_var - 4
... R_X86_64_TLSLD     ld_a   - 4
... R_X86_64_DTPOFF32  ld_a
... R_X86_64_DTPOFF32  ld_b
... R_X86_64_DTPOFF32  ld_a
... R_X86_64_DTPOFF32  ld_b
```

Run BOLT:

```sh
llvm-bolt libGcc.so -o libGcc.so.bolt

BOLT-WARNING: Failed to analyze 6 relocations   # 1x TLSGD + 1x TLSLD + 4x DTPOFF32

llvm-bolt libClang.so -o libClang.so.bolt

BOLT-WARNING: Failed to analyze 7 relocations   # 2x TLSGD + 1x TLSLD + 4x DTPOFF32
```

### Expected
No warning - GD/LD TLS relocations should be skipped just like the IE/LE TLS relocations already are.

## Proposed fix

Add the GD/LD model relocations to `isTLSX86()`, mirroring the other targets:

```cpp
static bool isTLSX86(uint32_t Type) {
  switch (Type) {
  default:
    return false;
  case ELF::R_X86_64_TPOFF32:    
  case ELF::R_X86_64_TPOFF64:    
  case ELF::R_X86_64_GOTTPOFF:   
  case ELF::R_X86_64_TLSGD:      // PC relative offset to GD GOT
  case ELF::R_X86_64_TLSLD:      // PC relative offset to LD GOT
  case ELF::R_X86_64_DTPOFF32:   // Offset in TLS block
  case ELF::R_X86_64_DTPOFF64:   // Offset in TLS block - > for completeness/consistency, see Reference 1
  case ELF::R_X86_64_DTPMOD64:   // Id of module containing symbol -> for completeness/consistency, see Reference 1
    return true;
  }
}
```

(Optionally, it may be worth rethinking to also cover other types defined under `llvm-project/llvm/include/llvm/BinaryFormat/ELFRelocs/x86_64.def`)

## References

1. TLS access models (GD / LD / IE / LE) and the relocations each one emits: Ulrich Drepper, *ELF Handling For Thread-Local Storage* - https://www.akkadia.org/drepper/tls.pdf §6.5 New x86-64 ELF Definitions

2. Relocation type numbers/definitions: *System V ABI — AMD64 psABI* - https://raw.githubusercontent.com/wiki/hjl-tools/x86-psABI/x86-64-psABI-1.0.pdf (Table 4.9 "Relocation Types", TLS entries).

3. llvm-project/llvm/include/llvm/BinaryFormat/ELFRelocs/x86_64.def - ideally BOLT would support all currently-unsupported types from that file
